.github/workflows/publish.yaml

```diff
@@ -0,0 +1,34 @@
+name: Publish
+
+on:
+  push:
+    branches:
+      - master
+      - main
+      - track/**
+  pull_request:
+    branches:
+      - master
+      - main
+      - track/**
+
+jobs:
+  publish-charm:
+    name: Publish Charm
+    runs-on: ubuntu-latest
+    # Only publish to charmhub if we are pushing to a special branch or running PRs from something named `branch/*`
+    if: (github.event_name == 'push') ||  (startsWith( github.head_ref, 'branch/' ))
```

Comment:
we **could** have this check, but on the other hand, we could drop it as the `branch/` part now is optional to release to charmhub branches.

https://github.com/canonical/charming-actions/blob/b01573cbc75f365ff02b5925ffc2d21f76246d93/src/services/ref/ref.ts#L44-L63

Reply:
Thanks @simskij. Do you think we should include some checks that are present e.g. [in prometheus-k8s charm](https://github.com/canonical/prometheus-k8s-operator/blob/main/.github/workflows/release-edge.yaml#L109)?

Reply:
That would make sense. We usually don't want to release something that we know is broken 👍🏼

Reply:
For context, this line was added back when the upload action would exit with a failure code in any cases that it shouldn't push (like say if we just updated a PR that was in a git branched named like `something-other-than-branch/`).  Without this `if` statement, cases that **correctly should not publish anything** would show up in our github checks as an :x: and look like a failure.  

Are there any cases with the current action now which we want to avoid publishing?  I haven't tracked how the logic evolved.  My only concern about editing this is that we ensure the github CI summary shows an intuitive message, eg:
* :heavy_check_mark: if it published successfully
* :x: if it failed to do whatever was expected (build failure, etc)
* (something) if we *should not* publish **and** that we actually *do not* publish  
  * if this is not relevant in current action, we can ignore
  * If this is still a relevant case, I argue this should not give a :heavy_check_mark: status as seeing ":heavy_check_mark: Publish Charm" in the CI summary implies that we **published** something, when instead we've correctly **decided not to publish anything**
  * Similar argument against giving status of :x: - it implies something went wrong, rather than we correctly decided not to publish anything

Reply:
Regarding adding checks, I like adding checks that block publishing to main.  But what about publishing to branches?  I think there's good reason to publish to a branch even if linting/unit/integration tests fail - often we're debugging multiple things at once and it would be nice to have a branch available during that process.